core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java

```diff
@@ -4102,6 +4102,43 @@ public static List reverse(List list) {
     return list;
   }
 
+  /** SQL {@code ARRAY_TO_STRING(array, delimiter)} function. */
+  public static String arrayToString(List list, String delimiter) {
+    return arrayToString(list, delimiter, null);
+  }
+
+  /** SQL {@code ARRAY_TO_STRING(array, delimiter, nullText)} function. */
+  public static String arrayToString(List list, String delimiter, @Nullable String nullText) {
+    StringBuilder sb = new StringBuilder();
+    for (Object item : list) {
+      String str;
+      if (item == null) {
+        str = null;
+      } else if (item instanceof String) {
+        str = (String) item;
+      } else if (item instanceof ByteString) {
+        str = item.toString();
+      } else {
+        throw new IllegalStateException(
+            "arrayToString supports only String or ByteString, but got "
+                + item.getClass().getName());
+      }
+      if (str == null) {
+        if (nullText == null) {
+          continue;
+        } else {
+          str = nullText;
+        }
+      }
+
+      if (sb.length() > 0) {
```

Comment:
Sorry @zoudan , one last detail (rather an edge case): what is the expected behavior if the item (`str` in this point) is the empty string? Will it be appended like any other value with the corresponding delimiter before the precious value? 
For example:
```
array_to_string(array['', 'b'], '-') => '-b'    ???
array_to_string(array['', ''], '-') => '-'    ???
```
If that is the case, I have the impression that this condition is wrong and we would need to find another mechanism, e.g.
```
boolean first = true;
for (Object item : list) {
  ...
  if (!first) {
    sb.append(delimiter);
  }
  sb.append(str);
  first = false;
} 
```
wdyt?

Reply:
@rubenada Thank you for your careful observation. Yes, this is the expected behavior:
`array_to_string(array['', 'b'], '-') => '-b'`
`array_to_string(array['', ''], '-') => '-' `
I have updated my code and also squashed the commits.

Reply:
Thanks @zoudan ! I saw that you also added some unit tests for this scenario, that's great.
I'll merge as soon as the CI checks are finished